lib/postgres.go

```diff
@@ -0,0 +1,1414 @@
+		}
+		if err := postgres.flushLikes(tx, view); err != nil {
+			return err
+		}
+		if err := postgres.flushFollows(tx, view); err != nil {
+			return err
+		}
+		if err := postgres.flushDiamonds(tx, view); err != nil {
+			return err
+		}
+		if err := postgres.flushMessages(tx, view); err != nil {
+			return err
+		}
+		if err := postgres.flushCreatorCoinBalances(tx, view); err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
+
+func (postgres *Postgres) flushUtxos(tx *pg.Tx, view *UtxoView) error {
+	var outputs []*PGTransactionOutput
+	for utxoKeyIter, utxoEntry := range view.UtxoKeyToUtxoEntry {
+		// Making a copy of the iterator is required
+		utxoKey := utxoKeyIter
+		outputs = append(outputs, &PGTransactionOutput{
+			OutputHash:  &utxoKey.TxID,
+			OutputIndex: utxoKey.Index,
+			OutputType:  utxoEntry.UtxoType,
+			PublicKey:   utxoEntry.PublicKey,
+			AmountNanos: utxoEntry.AmountNanos,
+			Spent:       utxoEntry.isSpent,
+		})
+	}
+
+	_, err := tx.Model(&outputs).WherePK().OnConflict("(output_hash, output_index) DO UPDATE").Insert()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (postgres *Postgres) flushProfiles(tx *pg.Tx, view *UtxoView) error {
+	var insertProfiles []*PGProfile
+	var deleteProfiles []*PGProfile
+	for pkidIter, profileEntry := range view.ProfilePKIDToProfileEntry {
+		// We don't have to do actually delete these profile entries. The DO UPDATE clause takes care of these changes.
+		if profileEntry.isDeleted {
+			glog.Infof("DELETED PROFILE %v %v", profileEntry.PublicKey, profileEntry.Username)
+			continue
+		}
+
+		// Making a copy of the iterator is required
+		pkid := pkidIter
+		profile := &PGProfile{
+			PKID:                    &pkid,
+			PublicKey:               NewPublicKey(profileEntry.PublicKey),
+			Username:                string(profileEntry.Username),
+			Description:             string(profileEntry.Description),
+			ProfilePic:              profileEntry.ProfilePic,
+			CreatorBasisPoints:      profileEntry.CreatorBasisPoints,
+			BitCloutLockedNanos:     profileEntry.BitCloutLockedNanos,
+			NumberOfHolders:         profileEntry.NumberOfHolders,
+			CoinsInCirculationNanos: profileEntry.CoinsInCirculationNanos,
+			CoinWatermarkNanos:      profileEntry.CoinWatermarkNanos,
+		}
+
+		if profileEntry.isDeleted {
+			deleteProfiles = append(deleteProfiles, profile)
+		} else {
+			insertProfiles = append(insertProfiles, profile)
+		}
+	}
+
+	if len(insertProfiles) > 0 {
+		_, err := tx.Model(&insertProfiles).WherePK().OnConflict("(pkid) DO UPDATE").Returning("NULL").Insert()
+		if err != nil {
+			return err
+		}
+	}
+
+	if len(deleteProfiles) > 0 {
+		_, err := tx.Model(&deleteProfiles).Returning("NULL").Delete()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (postgres *Postgres) flushPosts(tx *pg.Tx, view *UtxoView) error {
+	var insertPosts []*PGPost
+	var deletePosts []*PGPost
+	for _, postEntry := range view.PostHashToPostEntry {
+		post := &PGPost{
```

Comment:
Question: Is there anything stopping us from replacing PostEntry with PGPost, and similarly for the other structs? I think some fields are irrelevant, like isDeleted. But I think we can define those irrelevant fields as lowercase camelcase in the PGPost. 

I think we should merge as-is, but just a thought on something we could do later to minimize duplication.

Reply:
The goal is to eventually migrate from `PostEntry` to `PGPost` and delete PostEntry entirely once we're fully moved off of badger. Baby steps are needed as `**Entry` objects are everywhere, especially in backend, and doing everything at once is much more error prone.